File: 07/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <limits>


// [fixme] change std::string to std::size_t and use std::hash.
using Stipulation = std::pair<std::string, unsigned>;
using Rule = std::vector<Stipulation>;
using RuleMap = std::unordered_map<std::string, Rule>;


auto
findCount(Rule const& rule, std::string const& type) {
  return std::find_if(
    std::begin(rule),
    std::end(rule),
    [&type] (auto const& pair) {
      return pair.first == type;
    }
  );
}


std::string
readBagType(std::ifstream& input) {
  // Seems most bag names are at most 15 chars, so short string optimisation
  // hopefully means this isn't horribly inefficient.
  std::string adjective;
  std::string colour;
  input >> std::ws >> adjective;
  adjective += input.get();
  input >> colour;
  return adjective + colour;
}


template <typename InputStream>
InputStream&
operator>>(InputStream& input, std::pair<std::string, Rule>& node) {
  // <type>
  std::string ruleType = readBagType(input);

  // "bags contain"
  // constexpr auto offset = std::string(" bags contain ").size();
  constexpr int offset = 14;
  input.seekg(offset, std::ios_base::cur);

  Rule contents;
  char const next = input.peek();
  if ('1' <= next && next <= '9') {
    for (bool moreEntries = input.good(); moreEntries && input.good(); ) {
      // <count>
      unsigned count;
      input >> count >> std::ws;
      // <type>
      std::string type = readBagType(input);
      contents.emplace_back(std::move(type), count);
      input >> type; // bag/bags
      moreEntries = input.peek() != '\n';
    }
  }
  else {
    input.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); 
  }

  std::swap(node.first, ruleType);
  std::swap(node.second, contents);
  return input;
}


void
findAncestors(
  RuleMap const& rules,
  std::unordered_set<std::string>* ancestors,
  std::string const& type
) {
  for (auto const& node: rules) {
    auto const it = findCount(node.second, type);
    if (it != node.second.end()) {
      ancestors->emplace(node.first);
      findAncestors(rules, ancestors, node.first);
    }
  }
}


unsigned
countDescendants(
  RuleMap const& rules,
  std::string const& type,
  unsigned mult = 1
) {
  auto const node = rules.find(type);
  if (node == rules.end()) {
    return 0;
  }

  unsigned sum = 0;
  for (auto const& stipulation: node->second) {
    unsigned const count = stipulation.second;
    unsigned const descendantCount = countDescendants(
      rules, stipulation.first, count);
    sum += mult * (count + descendantCount);
  }
  return sum;
}


int main(int const argc, char const* const* const argv) {
  std::ifstream ifs(argc < 2 ? DEFAULT_INPUT_FILE : argv[1]);
  RuleMap rules;
  while (ifs.good() && ifs.peek() != std::fstream::traits_type::eof()) {
    std::pair<std::string, Rule> node;
    ifs >> node;
    rules.emplace(std::move(node));
  }

  std::unordered_set<std::string> ancestors;
  findAncestors(rules, &ancestors, "shiny gold");
  std::cout << "Part one: " << ancestors.size() << std::endl;
  std::cout << "Part two: " << countDescendants(rules, "shiny gold")
    << std::endl;
  return 0;
}

File: 07/main.js
#!/usr/bin/env node
const typeRegex = /^(.*) bags contain (.*)$/;
const contentsRegex = /(\d+) ([^,]+) bags?(, )?/g;


const rules = require('fs')
  .readFileSync(
    process.argv[2] || require('path').join(__dirname, 'input.txt'),
    'utf8'
  )
  .trim()
  .split('\n')
  .map(rule => {
    const typeMatch = typeRegex.exec(rule);
    const type = typeMatch[1];
    const contentsStr = typeMatch[2];
    const contents = {};
    if (typeMatch[2] !== 'no other bags.') {
      let contentsMatch = contentsRegex.exec(contentsStr);
      do {
        contents[contentsMatch[2]] = +contentsMatch[1];
      } while ((contentsMatch = contentsRegex.exec(contentsStr)) !== null);
    }
    return {rule, type, contentsStr, contents};
  });

const ruleMap = rules
  .reduce((dict, node) => {
    (dict[node.type] || (dict[node.type] = [])).push(node);
    return dict;
  }, {});

const findAncestors = (type, ancestors = new Set()) => {
  const parents = rules.filter(rule => rule.contents[type] > 0);
  for (const parent of parents) {
    if (!ancestors.has(parent)) {
      ancestors.add(parent);
      findAncestors(parent.type, ancestors);
    }
  }
  return ancestors;
};

const countDescendants = (type, mult = 1) => {
  let sum = 0;
  const rules = ruleMap[type];
  for (const rule of rules) {
    for (const childType in rule.contents) {
      const count = rule.contents[childType];
      sum += mult * (
        count + countDescendants(childType, rule.contents[childType])
      );
    }
  }
  return sum;
};


console.log(
  'Part one:', findAncestors('shiny gold').size,
  '\nPart two:', countDescendants('shiny gold'),
);
module.exports = rules;
module.exports.findAncestors = findAncestors;
module.exports.countDescendants = countDescendants;
module.exports.ruleMap = ruleMap;
